crmsh/utils.py

```diff
@@ -3107,4 +3110,32 @@ def time_value_with_unit(time_value):
     Check if the time value contains unit
     """
     return re.search(r'^\d+[a-z]+$', time_value) is not None
+
+
+def readline_sort_decorator(func):
+    '''
+    Decorator to disable readline sort completion matches
+
+    Python readline module sort the completion result by
+    alphabetical order by default. To archive the custom
+    sort order, need to disable the sort_completion_matches.
+    '''
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        readline_path = ctypes.util.find_library("readline")
+        if readline_path:
+            # Inspired by
+            # https://stackoverflow.com/questions/31229708/python-gnu-readline-bindings-keep-my-sort-order
+            rl = ctypes.cdll.LoadLibrary(readline_path)
+            sort = ctypes.c_ulong.in_dll(rl, 'rl_sort_completion_matches')
+            sort.value = 0
+            result = func(*args, **kwargs)
+            # Restore the default sort order later,
+            # after the completion matches are done.
+            timer = threading.Timer(1, lambda: setattr(sort, 'value', 1))
```

Comment:
This looks unreliable. Cab we just disable sort at readline level globally, and control whether to sort in our completers?

Reply:
Can we pass a flag representing _whether to sort_ along together with completion candidates, and change option `sort` accordingly just before the completion starts?

Reply:
> Can we pass a flag representing _whether to sort_ along together with completion candidates, and change option `sort` accordingly just before the completion starts?

Not get your point yet.
Using this decorator means disabling the default readline sort behavior

Reply:
Let's forget the decorator. We are calling readline to get some input from user. So the most suitable way to control whether to sort the candidates, is to set the option just before we call readline, and restore it after readline returns. What we need to do is pass this option down where readline is called, which should be in the RPEL I think.

Reply:
> restore it after readline returns
Don't clear how to do that, so I reset it before calling completer again